libsolidity/formal/Predicate.cpp

```diff
@@ -214,19 +215,64 @@ string Predicate::formatSummaryCall(
 	}
 
 	/// The signature of a function summary predicate is: summary(error, this, abiFunctions, cryptoFunctions, txData, preBlockChainState, preStateVars, preInputVars, postBlockchainState, postStateVars, postInputVars, outputVars).
-	/// Here we are interested in preInputVars to format the function call,
-	/// and in txData to retrieve `msg.value`.
+	/// Here we are interested in preInputVars to format the function call.
 
-	string value;
-	if (auto v = readTxVars(_args.at(4)).at("msg.value"))
+	string txModel;
+
+	if (_appendTxVars)
 	{
-		bigint x(*v);
-		if (x > 0)
-			value = "{ value: " + *v + " }";
+		set<string> txVars;
+		if (isFunctionSummary() && programFunction()->isPayable())
+			txVars.insert("msg.value");
+		else if (isConstructorSummary())
+		{
+			auto fun = programFunction();
+			if (fun && fun->isPayable())
+				txVars.insert("msg.value");
+		}
+
+		struct TxVarsVisitor: public ASTConstVisitor
+		{
+			bool visit(MemberAccess const& _memberAccess)
+			{
+				Expression const* memberExpr = SMTEncoder::innermostTuple(_memberAccess.expression());
+
+				auto const& exprType = memberExpr->annotation().type;
+				solAssert(exprType, "");
+				if (exprType->category() == Type::Category::Magic)
+				{
+					if (auto const* identifier = dynamic_cast<Identifier const*>(memberExpr))
+					{
+						auto const& name = identifier->name();
+						if (name == "block" || name == "msg" || name == "tx")
+							txVars.insert(name + "." + _memberAccess.memberName());
```

Comment:
Does this handle stuff like `(msg).value`? Seems to be valid syntax.

Reply:
Or how about this :)
```solidity
(true ? msg : msg).value;
```

Reply:
Fortunately something like `[msg, tx, block][i].value` is an error though :)

Reply:
sigh...

Reply:
This case was actually buggy already in develop, nice find :)
Fixed in https://github.com/ethereum/solidity/pull/12036